packages/bitcore-lib/lib/transaction/sighashschnorr.js

```diff
@@ -0,0 +1,231 @@
+'use strict';
+
+/* jshint maxparams:5 */
+
+const Signature = require('../crypto/signature');
+const BufferWriter = require('../encoding/bufferwriter');
+const Hash = require('../crypto/hash');
+const Schnorr = require('../crypto/schnorr');
+const $ = require('../util/preconditions');
+const TaggedHash = require('../crypto/taggedhash');
+const PrivateKey = require('../privatekey');
+
+/**
+ * Returns a buffer of length 32 bytes with the hash that needs to be signed
+ * for witness v1 programs as defined by:
+ * https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+ *
+ * @name Signing.sighash
+ * @param {Transaction} transaction the transaction to sign
+ * @param {Number} sighashType the type of the hash
+ * @param {Number} inputNumber the input index for the signature
+ * @param {Number} sigversion Taproot or Tapscript version number
+ * @param {Object} execdata Object with directives and data for creating the signature hash
+ */
+function _signatureHash(transaction, sighashType, inputNumber, sigversion, execdata) {
+  let extFlag, keyVersion;
+
+  switch (sigversion) {
+    case Signature.Version.TAPROOT:
+      extFlag = 0;
+      // keyVersion is not used and left uninitialized.
+      break;
+    case Signature.Version.TAPSCRIPT:
+      extFlag = 1;
+      // keyVersion must be 0 for now, representing the current version of
+      // 32-byte public keys in the tapscript signature opcode execution.
+      // An upgradable public key version (with a size not 32-byte) may
+      // request a different keyVersion with a new sigversion.
+      keyVersion = 0;
+      break;
+    default:
+      return false;
+  }
+  $.checkArgument(inputNumber < transaction.inputs.length, 'inputNumber is greater than number of inputs');
+
+  const ss = TaggedHash.TAPSIGHASH;
+
+  // Epoch
+  ss.writeUInt8(0);
+
+  // Hash type
+  const outputType = (sighashType == Signature.SIGHASH_DEFAULT) ? Signature.SIGHASH_ALL : (sighashType & Signature.SIGHASH_OUTPUT_MASK); // Default (no sighash byte) is equivalent to SIGHASH_ALL
+  const inputType = sighashType & Signature.SIGHASH_INPUT_MASK;
+  if (!(sighashType <= 0x03 || (sighashType >= 0x81 && sighashType <= 0x83))) {
+    return false;
+  }
+  ss.writeUInt8(sighashType);
+
+  // Transaction level data
+  ss.writeInt32LE(transaction.version);
+  ss.writeUInt32LE(transaction.nLockTime);
+  if (inputType !== Signature.SIGHASH_ANYONECANPAY) {
+    const prevoutsBW = new BufferWriter();
+    const spentAmountsBW = new BufferWriter();
+    const spentScriptsBW = new BufferWriter();
+    const sequencesBW = new BufferWriter();
+
+    for(let vin of transaction.inputs) {
+      prevoutsBW.writeReverse(vin.prevTxId);
+      prevoutsBW.writeInt32LE(vin.outputIndex);
+
+      spentAmountsBW.writeUInt64LEBN(vin.output._satoshisBN);
+
+      spentScriptsBW.writeUInt8(vin.output.script.toBuffer().length);
```

Comment:
Could assign vin.output.script.toBuffer() so you don't call toBuffer() twice.
`
const scriptBuffer = vin.output.script.toBuffer();
spentScriptsBW.writeUInt8(scriptBuffer.length);
spentScriptsBW.write(scriptBuffer);
`